enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,235 @@
+  between user logins
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+I am an administrator of an OpenShift cluster that uses identities from
+my company's identity provider. The users of the identity provider are
+organized in a complex and well-though-out group structure. I would like
+OpenShift to respect the group structure so that I can assign privileges
+to the users on OpenShift side according to their group membership.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### OIDC Group Retrieval
+
+OIDC core specification does not mention a standard `id_token` claim that
+would serve for user group storage and so none of the standard scopes define
+accessing it. That means retrieving the groups will very likely always require
+the user to configure `extraScopes` in their identity provider configuration
+(oauth.config.openshift.io/cluster).
+
+A possible follow-up implementation might include being able to
+[request specific claims](https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims)
+from the UserInfo endpoint or in an `id_token`, which would allow for a more fine-grained
+configuration to reach the group claim.
+
+#### API
+
+The `OpenIDClaims` config API structure shall be expanded by a new `Groups`
+field:
+
+```golang
+// OpenIDClaims contains a list of OpenID claims to use when authenticating with an OpenID identity provider
+type OpenIDClaims struct {
+    ...
+
+    // groups is the list of claims whose values should be converted into OpenShift groups
+    // that will include the user
+    // +optional
+    Groups []string `json:"groups,omitempty"`
+}
+```
+
+The group object should be annotated with `oauth.openshift.io/idpname: <idpname>`
+where `<idpname>` marks the identity provider name from the configuration. It should
+also get an `oauth.openshift.io/generated: true` if it hasn't previously existed.
+
+#### oauth-server
+
+When the OIDC provider has group synchronization configured, the oauth-server
+must attempt to retrieve the group claim or claims from the `id_token` it
+obtains from the provider after a successful authentication, and should do
+the same for the `id_token` retrieved from the OIDC's UserInfo endpoint
+should the OIDC be exposing this endpoint in its OIDC discovery data.
+
+Once the server has the user's groups information, it will look up the
+OpenShift groups the user appears in, and it will either add to/remove from/create
+the necessary groups for the user. The user should only be removed from the groups
+that were created for that specific identity provider (as per the annotation
+from the [API section](#api)). If such a group ends up being empty, it should
+be removed if it contains the `oauth.openshift.io/generated: true` annotation.
+
+When a group that would otherwise be created on sync already exists, the
+oauth-server should only add the user among the group members and add an
+`oauth.openshift.io/idpname` annotation to mark it was created for the
```

Comment:
could this be flapping if we have multiple identity providers configured?

Reply:
hmmm, good point, I think it I'll have to actually include the IdP name in the annotation itself